File: src/Apps/ArcaneKingdom/Unity/Assets/_Project/Scripts/Domain/Battle/BattleState.cs
#nullable enable
using System;
using System.Collections.Generic;
using ArcaneKingdom.Domain.Cards;

namespace ArcaneKingdom.Domain.Battle
{
    public enum BattlePhase
    {
        Setup = 0,
        PlayerTurn = 1,
        EnemyTurn = 2,
        TurnEnd = 3,
        Settlement = 4
    }

    public enum BattleResult
    {
        Undecided = 0,
        PlayerWins = 1,
        EnemyWins = 2,
        Draw = 3
    }

    /// <summary>
    /// Deterministischer Battle-State. Wird vom BattleEngine fortgeschritten.
    /// </summary>
    [Serializable]
    public sealed class BattleState
    {
        public int Seed { get; }
        public int CurrentTurn { get; set; }
        public int PlayerHeroHp { get; set; }
        public int EnemyHeroHp { get; set; }
        /// <summary>Maximales HP des Spieler-Helden (fuer 50%-Schwellen-Berechnung).</summary>
        public int PlayerHeroMaxHp { get; set; }
        /// <summary>Maximales HP des Gegner-Helden (fuer Boss-Phasen-Trigger).</summary>
        public int EnemyHeroMaxHp { get; set; }
        /// <summary>Verfuegbare Mana-Orbs des Spielers in der laufenden Runde (Designplan v3 Kap. 7.3: 3/Runde, 1 pro Karte).</summary>
        public int PlayerMana { get; set; }
        public int EnemyMana { get; set; }
        /// <summary>Mana-Orbs pro Runde (Reset-Wert). Spec: konstant 3, kein Anstieg ueber Runden.</summary>
        public int PlayerMaxMana { get; set; }
        public int EnemyMaxMana { get; set; }

        /// <summary>
        /// Anzahl in der laufenden Runde bereits eingesetzter Karten je Seite.
        /// Designplan v3 Kap. 7.3: Karten mit COST &gt; 30 koennen nur eingesetzt werden,
        /// wenn in diesem Zug noch nichts anderes gespielt wurde.
        /// </summary>
        public int PlayerCardsPlayedThisTurn { get; set; }
        public int EnemyCardsPlayedThisTurn { get; set; }

        /// <summary>
        /// Stat-Multiplier fuer Gegner-Karten nach Schwierigkeit (Spielplan v5 Kap. 8.3:
        /// Classic 1.0 / Amateur 1.25 / Profi 1.6 / Gott 2.2). Wird beim Einsetzen jeder
        /// Gegner-Karte auf ATK/HP angewandt (nicht auf das Spieler-Feld).
        /// </summary>
        public float EnemyStatMultiplier { get; set; } = 1.0f;

        /// <summary>
        /// Markiert den Kampf als Boss-Encounter (Mini-Boss = Node-Index 5, World-Boss = Node-Index 10).
        /// Aktiviert die Boss-Phase-2-Mechanik (Spielplan v5 Kap. 9.4).
        /// </summary>
        public bool IsBossEncounter { get; set; }

        /// <summary>
        /// Karten-Definitionen die als Verstaerkung in Boss-Phase-2 ins Feld kommen
        /// (2-3 starke Karten laut Plan). Wird beim Setup gefuellt.
        /// </summary>
        public List<string> BossPhase2ReinforcementCardIds { get; } = new();

        /// <summary>
        /// Passive Faehigkeit, die in Boss-Phase-2 aktiviert wird
        /// (z.B. "Alle Karten +200 ATK"). Lokalisierungs-Key oder Beschreibung.
        /// </summary>
        public string? BossPhase2PassiveKey { get; set; }

        /// <summary>True wenn Boss-Phase-2 bereits ausgeloest wurde (verhindert doppelten Trigger).</summary>
        public bool BossPhase2Active { get; set; }
        public List<CardFieldSlot> PlayerField { get; }      // max. 5
        public List<CardFieldSlot> EnemyField { get; }       // max. 5
        public List<string> PlayerHand { get; }              // InstanceIds
        public List<string> EnemyHand { get; }
        public Queue<string> PlayerDeckQueue { get; }
        public Queue<string> EnemyDeckQueue { get; }
        public BattlePhase Phase { get; set; }
        public BattleResult Result { get; set; }

        /// <summary>
        /// Helden-Passiv-Kontext pro Seite (Designplan v4 Kap. 2.1). Wird beim Setup gefüllt.
        /// </summary>
        public HeroPassivContext? PlayerHeroPassiv { get; set; }
        public HeroPassivContext? EnemyHeroPassiv { get; set; }

        /// <summary>
        /// Aggregierte Deck-Runen-Boni (Spielplan v5 Kap. 7.2). Vom BattleBootstrap vorberechnet,
        /// in Setup/PlayCard/EndTurn angewandt. Null = keine Runen. NICHT im Replay-Snapshot
        /// (Serializer ist DTO-basiert) — Boni werden in HeroHp/Field-Werte eingebrannt.
        /// EnemyRuneLoadout ist Phase 1 immer null (symmetrisch fuer spaeteren PvP).
        /// </summary>
        public ArcaneKingdom.Domain.Runes.RuneLoadout? PlayerRuneLoadout { get; set; }
        public ArcaneKingdom.Domain.Runes.RuneLoadout? EnemyRuneLoadout { get; set; }

        /// <summary>
        /// Karten-Persönlichkeit-Events (Designplan v4 Kap. 8) für UI/Animation/Replay.
        /// </summary>
        public List<BattleEvent> Events { get; }

        public BattleState(int seed, int playerHeroHp, int enemyHeroHp)
        {
            Seed = seed;
            PlayerHeroHp = playerHeroHp;
            EnemyHeroHp = enemyHeroHp;
            PlayerHeroMaxHp = playerHeroHp;
            EnemyHeroMaxHp = enemyHeroHp;
            PlayerMana = 3;
            EnemyMana = 3;
            PlayerMaxMana = 3;
            EnemyMaxMana = 3;
            PlayerField = new List<CardFieldSlot>(5);
            EnemyField = new List<CardFieldSlot>(5);
            PlayerHand = new List<string>(8);
            EnemyHand = new List<string>(8);
            PlayerDeckQueue = new Queue<string>();
            EnemyDeckQueue = new Queue<string>();
            Phase = BattlePhase.Setup;
            Result = BattleResult.Undecided;
            CurrentTurn = 1;
            Events = new List<BattleEvent>(64);
        }
    }

    /// <summary>
    /// Karten-Feld-Slot mit aktuellen Werten (ATK/HP können Buffs/Debuffs erlitten haben).
    /// </summary>
    [Serializable]
    public sealed class CardFieldSlot
    {
        public string CardInstanceId { get; }
        public int CurrentAttack { get; set; }
        public int CurrentHealth { get; set; }
        public int MaxHealth { get; set; }              // fuer Vorgaben wie "+5% HP" beim KoeniglicheAura
        public int TurnsUntilSpecial { get; set; }

        /// <summary>
        /// Aktive Status-Effekte auf dieser Karte (Designplan v4 Kap. 3.4):
        /// Schlaf, Stille, Einfrierung, Betaeubung, Vergiftung, Verbrennung, Verlangsamung, Verwurzelung.
        /// </summary>
        public List<StatusEffect> StatusEffects { get; } = new();

        public CardFieldSlot(string cardInstanceId, int currentAttack, int currentHealth, int turnsUntilSpecial)
        {
            CardInstanceId = cardInstanceId;
            CurrentAttack = currentAttack;
            CurrentHealth = currentHealth;
            MaxHealth = currentHealth;
            TurnsUntilSpecial = turnsUntilSpecial;
        }
    }
}

File: src/Apps/ArcaneKingdom/Server/CloudFunctions/src/battle/engine/battleState.ts
// Battle-State + CardFieldSlot + HeroPassivContext + BattleEvent — 1:1-Portierung von
// Unity/Assets/_Project/Scripts/Domain/Battle/{BattleState,HeroPassivContext,BattleEvent}.cs.
//
// Determinismus-relevant:
//   - PlayerDeckQueue/EnemyDeckQueue sind FIFO-Queues (C# Queue<string>): enqueue() haengt an,
//     dequeue() nimmt das vorderste Element. Wir bilden das mit einem Array + Kopf-Index nach,
//     damit shift() (O(n)) nicht doch eine andere Reihenfolge erzeugt — Reihenfolge ist hier kritisch.
//   - MaxHealth wird im Ctor von CardFieldSlot auf CurrentHealth gesetzt (wie C#).

import {
  BattlePhase,
  BattleResult,
  BattleEventType,
  HeroFaehigkeitsTyp,
} from "./types";
import { StatusEffect } from "./statusEffect";

/** Karten-Feld-Slot mit aktuellen Kampf-Werten. */
export class CardFieldSlot {
  readonly cardInstanceId: string;
  currentAttack: number;
  currentHealth: number;
  maxHealth: number;
  turnsUntilSpecial: number;
  readonly statusEffects: StatusEffect[] = [];

  constructor(cardInstanceId: string, currentAttack: number, currentHealth: number, turnsUntilSpecial: number) {
    this.cardInstanceId = cardInstanceId;
    this.currentAttack = currentAttack;
    this.currentHealth = currentHealth;
    this.maxHealth = currentHealth; // C#-Ctor: MaxHealth = currentHealth
    this.turnsUntilSpecial = turnsUntilSpecial;
  }
}

/** Vorberechnete Helden-Passiv-Parameter pro Kampfseite. */
export class HeroPassivContext {
  readonly passivType: HeroFaehigkeitsTyp;
  readonly magnitude: number;
  beastSpiritCountInDeck: number;
  firstCardThisTurnPlayed: boolean;
  divineBlessingsRemaining: number;

  constructor(passivType: HeroFaehigkeitsTyp, magnitude: number, beastSpiritCountInDeck = 0) {
    this.passivType = passivType;
    this.magnitude = magnitude;
    this.beastSpiritCountInDeck = beastSpiritCountInDeck;
    this.firstCardThisTurnPlayed = false;
    // Wie C#: nur GoettlicherSegen startet mit Rettungen, sonst 0.
    this.divineBlessingsRemaining = passivType === HeroFaehigkeitsTyp.GoettlicherSegen ? magnitude : 0;
  }
}

/** Aufzeichnung eines Battle-Events fuer Replay/Reporting. */
export class BattleEvent {
  readonly eventType: BattleEventType;
  readonly turn: number;
  readonly forPlayer: boolean;
  readonly cardInstanceId: string | null;
  readonly cardDefinitionId: string | null;
  readonly localizationKey: string | null;
  readonly partnerCardId: string | null;
  readonly magnitude: number;

  constructor(
    eventType: BattleEventType,
    turn: number,
    forPlayer: boolean,
    opts: {
      cardInstanceId?: string | null;
      cardDefinitionId?: string | null;
      localizationKey?: string | null;
      partnerCardId?: string | null;
      magnitude?: number;
    } = {}
  ) {
    this.eventType = eventType;
    this.turn = turn;
    this.forPlayer = forPlayer;
    this.cardInstanceId = opts.cardInstanceId ?? null;
    this.cardDefinitionId = opts.cardDefinitionId ?? null;
    this.localizationKey = opts.localizationKey ?? null;
    this.partnerCardId = opts.partnerCardId ?? null;
    this.magnitude = opts.magnitude ?? 0;
  }
}

/**
 * FIFO-Queue ueber einem Array mit wanderndem Kopf-Index — bildet C# Queue<string> exakt nach
 * (enqueue haengt an, dequeue liefert das vorderste Element).
 */
export class StringQueue {
  private readonly items: string[] = [];
  private head = 0;

  enqueue(value: string): void {
    this.items.push(value);
  }

  dequeue(): string | undefined {
    if (this.head >= this.items.length) return undefined;
    return this.items[this.head++];
  }

  get count(): number {
    return this.items.length - this.head;
  }
}

/** Deterministischer Battle-State. Wird von der BattleEngine fortgeschritten. */
export class BattleState {
  readonly seed: number;
  currentTurn: number;
  playerHeroHp: number;
  enemyHeroHp: number;
  playerHeroMaxHp: number;
  enemyHeroMaxHp: number;
  playerMana: number;
  enemyMana: number;
  playerMaxMana: number;
  enemyMaxMana: number;
  playerCardsPlayedThisTurn: number;
  enemyCardsPlayedThisTurn: number;
  enemyStatMultiplier: number;
  isBossEncounter: boolean;
  readonly bossPhase2ReinforcementCardIds: string[] = [];
  bossPhase2PassiveKey: string | null;
  bossPhase2Active: boolean;
  readonly playerField: CardFieldSlot[] = [];
  readonly enemyField: CardFieldSlot[] = [];
  readonly playerHand: string[] = [];
  readonly enemyHand: string[] = [];
  readonly playerDeckQueue = new StringQueue();
  readonly enemyDeckQueue = new StringQueue();
  phase: BattlePhase;
  result: BattleResult;
  playerHeroPassiv: HeroPassivContext | null;
  enemyHeroPassiv: HeroPassivContext | null;
  readonly events: BattleEvent[] = [];

  constructor(seed: number, playerHeroHp: number, enemyHeroHp: number) {
    this.seed = seed;
    this.playerHeroHp = playerHeroHp;
    this.enemyHeroHp = enemyHeroHp;
    this.playerHeroMaxHp = playerHeroHp;
    this.enemyHeroMaxHp = enemyHeroHp;
    this.playerMana = 3;
    this.enemyMana = 3;
    this.playerMaxMana = 3;
    this.enemyMaxMana = 3;
    this.playerCardsPlayedThisTurn = 0;
    this.enemyCardsPlayedThisTurn = 0;
    this.enemyStatMultiplier = 1.0;
    this.isBossEncounter = false;
    this.bossPhase2PassiveKey = null;
    this.bossPhase2Active = false;
    this.phase = BattlePhase.Setup;
    this.result = BattleResult.Undecided;
    this.currentTurn = 1;
    this.playerHeroPassiv = null;
    this.enemyHeroPassiv = null;
  }
}
